File: src/bfsdfs/dfs/boj/Main_유기농배추_1012.java
package bfsdfs.dfs.boj;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main_유기농배추_1012 {

	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringBuilder sb = new StringBuilder();
	static int N, M, K;
	static boolean[][] map;
	static boolean[][] visited;

	static int count = 0;

	public static void main(String[] args) throws IOException {
		int T = Integer.parseInt(br.readLine());

		for (int t = 0; t < T; t++) {
			count = 0;
			String[] NMK = br.readLine().split(" ");

			M = Integer.parseInt(NMK[0]);
			N = Integer.parseInt(NMK[1]);
			K = Integer.parseInt(NMK[2]);
			map = new boolean[N][M];
			for (int k = 0; k < K; k++) {
				String[] XY = br.readLine().split(" ");
				int X = Integer.parseInt((XY[0]));
				int Y = Integer.parseInt((XY[1]));
				map[Y][X] = true;
			}

			visited = new boolean[N][M];

			for (int r = 0; r < N; r++) {
				for (int c = 0; c < M; c++) {
					if (!visited[r][c] && map[r][c]) {
						dfs(r, c);
						count++;
					}
				}
			}
			sb.append(count).append('\n');
		}
		System.out.println(sb);
	}

	static int[] dy = {0, 0, 1, -1};
	static int[] dx = {1, -1, 0, 0};

	private static void dfs(int r, int c) {
		visited[r][c] = true;

		for (int d = 0; d < 4; d++) {
			int ty = r + dy[d];
			int tx = c + dx[d];

			if (isValid(ty, tx) && !visited[ty][tx] && map[ty][tx]) {
				dfs(ty, tx);
			}
		}
	}

	private static boolean isValid(int r, int c) {
		return r >= 0 && r < N && c >= 0 && c < M;
	}

}

File: src/bfsdfs/dfs/boj/Main_유기농배추_1012.kt
package bfsdfs.dfs.boj

import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {

    val br = BufferedReader(InputStreamReader(System.`in`))
    val t = br.readLine().toInt()
    for (i in 1..t) {
        val mnk = br.readLine().split(" ")
        val m = mnk[0].toInt()
        val n = mnk[1].toInt()
        val k = mnk[2].toInt()

        val map = Array(n) { BooleanArray(m) }
        for (j in 1..k) {
            val xy = br.readLine().split(" ")
            val x = xy[0].toInt()
            val y = xy[1].toInt()
            map[y][x] = true
        }

        Solution(n, m, map).solve()
    }
}

class Solution(
        val n: Int,
        val m: Int,
        val map: Array<BooleanArray>
) {
    val visited: Array<BooleanArray> = Array(n) { BooleanArray(m) }

    fun solve() {
        var answer = 0

        for (r in 0 until n) {
            for (c in 0 until m) {
                if (!visited[r][c] && map[r][c]) {
                    dfs(r, c, 0)
                    answer++
                }
            }
        }
        println(answer)
    }

    val dy = arrayOf(0, 0, 1, -1)
    val dx = arrayOf(1, -1, 0, 0)

    fun dfs(r: Int, c: Int, cnt: Int) {
        visited[r][c] = true
        for (d in 0 until 4) {
            val ty = r + dy[d]
            val tx = c + dx[d]

            if (isValid(ty, tx) && !visited[ty][tx] && map[ty][tx]) {
                dfs(ty, tx, cnt + 1);
            }
        }
    }

    fun isValid(r: Int, c: Int): Boolean {
        return r >= 0 && r < n && c >= 0 && c < m
    }
}